lib/services/oauth/service.go

```diff
@@ -0,0 +1,121 @@
+package oauth
+
+import (
+	"github.com/RangelReale/osin"
+	"github.com/jmoiron/sqlx"
+)
+
+// Service implements the Oauth service, returning Oauth data from the DB
+type Service struct {
+	db *sqlx.DB
+}
+
+// AuthorizeID is the ID of an authorization object
+type AuthorizeID string
+
+// GetClientByID returns an osin.Client by its ID
+func (s *Service) GetClientByID(id int) (osin.Client, error) {
+	dbClient := Client{}
+	err := s.db.Get(&dbClient, "SELECT * FROM clients WHERE id = $1", id)
+	if err != nil {
+		return &osin.DefaultClient{}, err
+	}
+	return &osin.DefaultClient{
+		Id:          dbClient.ExternalID,
+		Secret:      dbClient.Secret,
+		RedirectUri: dbClient.RedirectURI,
+		UserData:    dbClient.UserData,
+	}, nil
+}
+
+// GetAuthorizationClientByID returns osin.AuthorizationData for a client
+// TODO: make canned erors
+func (s *Service) GetAuthorizationClientByID(id int) (osin.AuthorizeData, error) {
+	dbAuthorization := Authorization{}
+	err := s.db.Get(&dbAuthorization, "SELECT * FROM authorizations WHERE client_id = $1", id)
+	if err != nil {
+		return osin.AuthorizeData{}, err
+	}
+
+	client, err := s.GetClientByID(dbAuthorization.ClientID)
+	if err != nil {
+		return osin.AuthorizeData{}, err
+	}
+
+	return osin.AuthorizeData{
+		Client:      client,
+		Code:        dbAuthorization.Code,
+		ExpiresIn:   dbAuthorization.Expiration,
+		Scope:       dbAuthorization.Scope,
+		RedirectUri: dbAuthorization.RedirectURI,
+		UserData:    dbAuthorization.UserData,
+	}, nil
+}
+
+// CreateClient creates a client DB object
+func (s *Service) CreateClient(client osin.Client) error {
+	_, err := s.db.Exec(
+		`INSERT INTO clients (
+                        external_id,
+                        secret,
+                        redirect_uri,
+                        user_data) VALUES ($1, $2, $3, $4)`,
+		client.GetId(),
+		client.GetSecret(),
+		client.GetRedirectUri(),
+		client.GetUserData(),
+	)
+	return err
+}
+
+// CreateAuthorizationData creates a DB authorization object
+func (s *Service) CreateAuthorizationData(data osin.AuthorizeData) error {
```

Comment:
Should this return the newly created object? And ditto with CreateAccess

Reply:
I decided to not have it return the newly created object, because these methods just de/serialize `osin` objects, but if you feel that we should return the object, then I can do that.

Reply:
I think it depends on how we/osin end up using these things. If we don't need it for now, then no need to add it